rfc/redis-removal.md

```diff
@@ -0,0 +1,171 @@
+overhead of running Spinnaker, Redis needs to be deprecated and removed from
+being a persistence option in favor of using SQL as the baseline storage
+technology. This RFC documents a process and timeline for doing so.
+
+### Goals and Non-Goals
+
+Goals:
+
+*   Provide an inventory of Spinnaker services that still need SQL support added
+*   Define a process that all services can use for migrating datastores without
+    downtime
+*   Provide a timeline for the final switch to SQL, along with criteria to
+    decide if the timeline should be delayed
+
+Non-Goals:
+
+*   Debating if SQL is the best replacement default persistence backend to
+    standardize on
+*   Documenting and solving all technical issues for each service to migrate;
+    those will be handled separately whereas this document concerns itself
+    primarily with the larger processes and timeline
+*   Identification of all new processes for maintaining services on a SQL
+    backend; these will be handled separately
+*   Prescription of what SQL vendor to use: Current functionality has been built
+    using MySQL; support for other vendors will be addressed separately
+
+## Case Against Redis
+
+Redis was introduced into the project as an easy way to get started. It's
+phenomenally flexible, and a single, common data store simplifies the
+operational story of the large operational footprint that Spinnaker has.
+Further, it was useful in that its flexibility meant services could be rapidly
+iterated on, as no schema was needed from the start.
+
+Unfortunately, Redis is difficult to operationalize as a persistent datastore,
+which many of the services treat Redis as. Community members are often surprised
+when Redis restarts to find Spinnaker has lost in-flight work. Operating Redis
+beyond a cache is not a skill that many organizations have. Worse yet, the level
+of effort to support a highly-available, persistent story for Redis is very high
+and littered with edge cases. For example, when tested at Netflix, both Redis
+Cluster and Sentinel regularly encountered incorrect master failover behavior
+that would lead to corrupted clusters or lost data.
+
+Furthermore, developing applications on Redis is typically a foreign concept and
+offers a confusing touch point for anyone looking to get involved in
+contributing.
+
+Having a singular datastore for Spinnaker is still necessary, as it eases the
+burden of onboarding. Now that Spinnaker is no longer a young project, the need
+for fast service iteration has largely dissapated. SQL was chosen as the
+replacement because modern SQL is flexible and well understood.
+
+## Service Inventory
+
+Most of the larger services already have support for SQL.
+
+* Full support: ✔️
+* Partial support: ~
+* No support: ✗
+
+| | |
+|-|-|
+| Clouddriver | ✔️ |
+| Echo | ✗ |
+| Fiat | ✗ |
+| Front50 | ✔️ |
+| Gate | ~ |
+| Igor | ✗ |
+| Kayenta | ~ |
+| Keel | ✔️ |
+| Orca | ✔️ |
+| Rosco | ✗ |
+| Swabbie | ✗ |
+
+## Service Migration Process
```

Comment:
Any thoughts about testing processes for this? This feels like potentially the trickiest technical part, and also the part that is hardest to test, and also the part that is most likely to cause catastrophic issues if it goes wrong. So it makes me feel 😱

Maybe there is a tool that can round-trip the data from Redis to MySQL and back, and then verify that the data is completely unchanged. Then we could tell customers "if you can run this round-trip tool on your Redis database, you can trust that we'll migrate it safely"? Just a first idea, maybe not a great one.

Reply:
Testing tools would be good, yeah. We've done the other services with unit & integration tests, but for the most part did things yolo after a few practice runs in a lower environment.

The round-trip tool is interesting an interesting idea. It'd definitely give people more confidence into the migration process. I might give thing a try on one of the smaller services to see how it feels. 💯